secboot/secboot_tpm.go

```diff
@@ -309,7 +321,14 @@ func SealKeys(keys []SealKeyRequest, params *SealKeysParams) error {
 
 	if params.TPMProvision {
 		// Provision the TPM as late as possible
-		if err := tpmProvision(tpm, params.TPMLockoutAuthFile); err != nil {
+		existingLockoutAuth := params.TPMPartialReprovision
+		if err := tpmProvision(tpm, params.TPMLockoutAuthFile, existingLockoutAuth); err != nil {
+			return err
+		}
+	}
+	if params.TPMPartialReprovision {
+		logger.Debugf("releasing TPM NV resource with handle %v", params.PCRPolicyCounterHandle)
+		if err := tpmReleaseResources(tpm, tpm2.Handle(params.PCRPolicyCounterHandle)); err != nil {
```

Comment:
This code isn't atomic because it makes existing keys unrecoverable before creating new ones. As this code path needs to run for the fallback sealed keys (one of which is needed to unlock ubuntu-save), it creates a window where an interruption will result in data loss.

The process should seal new keys first (which implies creating a new PCR policy counter) and then erase counters associated with old keys once the system is bootable with the new keys.

Reply:
Sounds like this would need to be more complicated then, at least for ubuntu-save. Which means 2 more policy counter handles, a way to confirm those are in use after a boot. Then after the boot, we'd release the old ones. Next we either reeal the new keys using old handles (is that even possible? or should we seal them again?) and release the new handles. Otherwise if it's feasible to track or probe that a 2nd set of handles was used, we could stay on the alternative handles for the rest of device lifetime. Is this making any sense?

Reply:
seems we need two set of handles to use in round-robin fashion? it seems also that this requires splitting (re)provisioning, sealing and resetting resources as described above?

indeed, can we detect if a handle is in use?